### ISSUE TITLE
How to set a timeout for GoogleCredentials?

### ISSUE THREAD

**Issue by hiranya911** [googleapis/google-auth-library-java#356]
`GoogleCredentials` can be initialized with an optional `HttpTransportFactory`. But this doesn't provide any means of setting the connection and read timeout for the HTTP requests made by the credentials implementation. Is there a way to set these timeout values for `GoogleCredentials`?

More context: In the Google HTTP client, the recommended way to set timeouts is via an `HttpRequestInitializer` that gets passed into the `HttpRequestFactory`. But currently there appears to be no way to  specify a request initializer for `GoogleCredentials`. As a result, there's no simple way to configure the timeouts for the HTTP requests made by this implementation.

This is a somewhat critical issue, since `GoogleCredentials` get invoked before any other HTTP requests are made by the client. The default timeout settings of `GoogleCredentials` essentially mask and override any timeout values set in the client.

**Comment by hiranya911** [comment]
There's also some retry logic baked into the flow that is not configurable:

https://github.com/googleapis/google-auth-library-java/blob/f4f05be44ab073ee16d92a1647e74b8bb187a442/oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java#L404-L425

By default failing requests are retried up to 10 times with exponential backoff.

**Comment by codyoss** [comment]
I agree, there does not seem to be a easy way to do this today. Is your use case you would like to fail faster in your code? Although it is not ideal, I bet you could add timeouts in by wrapping `GoogleCredentials` as not a lot seems to be final.

**Comment by hiranya911** [comment]
Yes, fail fast would be one use case. 

> Although it is not ideal, I bet you could add timeouts in by wrapping GoogleCredentials as not a lot seems to be final.

Does that imply implementing some sort of a timer around the `GoogleCredentials.refreshAccessToken()` API? 

It appears there's no easy way to implement actual connection/socket timeouts since HTTP request logic is essentially buried in the `GoogleCredentials.refreshAccessToken()` method. Couple of solutions I can think of:
1. Make it possible to pass an `HttpRequestInitializer` into `GoogleCredentials`.
2. Expose an `initializeTokenRequest(HttpRequest req)` method from `GoogleCredentials` that users can override. Default implementation will do nothing.

**Comment by codyoss** [comment]
> Does that imply implementing some sort of a timer around the GoogleCredentials.refreshAccessToken() API?
That method needs to be implemented, its default impl from the base class is to throw an exception. Are you using GoogleCredentials directly or one of the other classes that implement it?

I will bring those possible solutions up with my team. Thank you for your input!

**Comment by hiranya911** [comment]
Ours is also a developer product: https://github.com/firebase/firebase-admin-java

Developers pass `GoogleCredentials` instances to our SDK. I believe most developers would use one of the factory methods to create credentials -- `GoogleCredentials.fromStream(...)`. The actual instance they get is something like `ServiceAccountCredentials` or `ApplicationDefaultCredentials`. 

It would be great if there was an API to wrap an existing `GoogleCredentials` instance with a timeout. This way our SDK can "inject" a timeout to the object passed in by the developer:

```
GoogleCredentials copy = credentials.withTimeout(connectTimeout, readTimeout);

// Or more generally
GoogleCredentials copy = credentials.withHttpRequestInitializer(initializer);
```

But we can also settle for a solution that would require developers to initialize `GoogleCredentials` with a timeout or a request initializer.

**Comment by anmol1vw13** [comment]
@hiranya911 @yoshi-automation Is there an update on this?

**Comment by stager0909-ray** [comment]
@hiranya911 @yoshi-automation Is there an update on this?

I am sending pushes using firebase-admin sdk.
The following domain could not be reached due to an internal network issue.
```
oauth2.googleapis.com
fcm.googleapis.com

```
When connecting to fcm.googleapis.com, the settings are as follows.
```
connectionTimeout = 3000
readTimeout = 30000
```

However, the setting value of the request to oauth2.googleapis.com in ServiceAccountCredentials.refreshToken() of google-auth-library-auth2-http-1.7.0 was the default.
```
connectionTimeout = 20000
readTimeout = 20000

```
ServiceAccountCredentials.refreshToken()
```
HttpRequestFactory requestFactory = transportFactory.create().createRequestFactory();
HttpRequest request = requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), content);

```
request has ApacheHttpTransport, and the settings in it are as follows.
connectionTimeout = 3000
readTimeout = 30000
However, request does not use the transport configuration, and uses the default value of 20 seconds.
![image](https://user-images.githubusercontent.com/39253019/195093714-ca715cd1-d0b1-42be-a8f4-2b88f7773da1.png)

I think this should be patched.

Do you have any patch plans?

**Comment by stager0909** [comment]
https://github.com/googleapis/google-auth-library-java/pull/1053

please check merge request
